packages/flutter/lib/src/material/segmented_button.dart

```diff
@@ -590,32 +619,25 @@ class SegmentedButtonState<T> extends State<SegmentedButton<T>> {
       );
     }
 
-    final OutlinedBorder resolvedEnabledBorder =
-        resolve<OutlinedBorder?>((ButtonStyle? style) => style?.shape, disabledState) ??
-        const RoundedRectangleBorder();
-    final OutlinedBorder resolvedDisabledBorder =
-        resolve<OutlinedBorder?>((ButtonStyle? style) => style?.shape, disabledState) ??
+    final OutlinedBorder effectiveBorder =
+        resolve<OutlinedBorder?>((ButtonStyle? style) => style?.shape) ??
         const RoundedRectangleBorder();
-    final BorderSide enabledSide =
-        resolve<BorderSide?>((ButtonStyle? style) => style?.side, enabledState) ?? BorderSide.none;
-    final BorderSide disabledSide =
-        resolve<BorderSide?>((ButtonStyle? style) => style?.side, disabledState) ?? BorderSide.none;
-    final OutlinedBorder enabledBorder = resolvedEnabledBorder.copyWith(side: enabledSide);
-    final OutlinedBorder disabledBorder = resolvedDisabledBorder.copyWith(side: disabledSide);
+    final BorderSide effectiveSide =
+        resolve<BorderSide?>((ButtonStyle? style) => style?.side) ?? BorderSide.none;
+    final OutlinedBorder enabledBorder = effectiveBorder.copyWith(side: effectiveSide);
+    final OutlinedBorder disabledBorder = effectiveBorder.copyWith(side: effectiveSide);
```

Comment:
Following my explanation above, this `disabledBorder` will never get the disabled value if a segment is disabled by setting `ButtonSegment.enabled` (like the screenshot below). So we might still need to use the original implementation to handle the disabled state.

![Screenshot 2025-04-08 at 9 10 54 PM](https://github.com/user-attachments/assets/0143d4d1-f5f7-4782-9ca1-38d918e1eb6c)